contrib/agent/src/main/java/io/opencensus/contrib/agent/AgentMain.java

```diff
@@ -62,14 +65,46 @@ public static void premain(String agentArgs, Instrumentation instrumentation) th
     instrumentation.appendToBootstrapClassLoaderSearch(
             new JarFile(Resources.getResourceAsTempFile("bootstrap.jar")));
 
+    checkLoadedByBootstrapClassloader(ContextManager.class);
+    checkLoadedByBootstrapClassloader(ContextStrategy.class);
+
     AgentBuilder agentBuilder = new AgentBuilder.Default()
             .disableClassFormatChanges()
             .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
             .with(new AgentBuilderListener())
             .ignore(none());
-    // TODO(stschmidt): Add instrumentation for context propagation.
+    agentBuilder = LazyLoaded.addGrpcContextPropagation(agentBuilder);
+    // TODO: Add more instrumentation, potentially introducing a plugin mechanism.
     agentBuilder.installOn(instrumentation);
 
     logger.info("Initialized.");
   }
+
+  private static void checkLoadedByBootstrapClassloader(Class<?> clazz) {
+    checkState(clazz.getClassLoader() == null,
+            "%s must be loaded by the bootstrap classloader",
+            clazz);
+  }
+
+  private static class LazyLoaded {
+
+    /**
+     * Adds automatic context propagation.
+     */
+    static AgentBuilder addGrpcContextPropagation(AgentBuilder agentBuilder) {
```

Comment:
I think we should not name it grpc, also an open issue to move context into a separate package which does not contain gRPC is in progress (discussions around that subject for the moment).